src/main/scala/is/hail/expr/ir/functions/ArrayFunctions.scala

```diff
@@ -191,16 +230,25 @@ object ArrayFunctions extends RegistryFunctions {
 
     registerIR("uniqueMaxIndex", TArray(tv("T")))(uniqueIndex(_, GT))
 
-    registerIR("[]", TArray(tv("T")), TInt32()) { (a, i) => ArrayRef(a, i) }
+    registerIR("[]", TArray(tv("T")), TInt32()) { (a, i) =>
+      ArrayRef(
+        a,
+        If(ApplyComparisonOp(LT(TInt32()), i, I32(0)),
+          ApplyBinaryPrimOp(Add(), ArrayLen(a), i),
+          i))
+    }
 
     registerIR("[:]", TArray(tv("T"))) { (a) => a }
 
     registerIR("[*:]", TArray(tv("T")), TInt32()) { (a, i) =>
       val idx = genUID()
       ArrayMap(
-        ArrayRange(If(ApplyComparisonOp(LT(TInt32()), i, I32(0)),
-          ApplyBinaryPrimOp(Add(), ArrayLen(a), i),
-          i),
+        ArrayRange(
+          If(ApplyComparisonOp(LT(TInt32()), i, I32(0)),
+            UtilFunctions.max(
+              ApplyBinaryPrimOp(Add(), ArrayLen(a), i),
```

Comment:
Why do you need `max` here? It's not in the original FunctionRegistry code. Did we not have the same behavior as Python before? If yes, then we should fix in the FunctionRegistry as well.

Reply:
Yeah, this is to match Python. I just checked the function registry, and after making the changes I realized that the old code is actually right, because of the behavior of Scala's `slice`.